climada/engine/test/test_impact_calc.py

```diff
@@ -0,0 +1,407 @@
+        self.hazard = create_autospec(HAZ)
+        self.hazard.get_mdr.return_value = sparse.csr_matrix(
+            [[0.0, 0.5, -1.0], [1.0, 2.0, 1.0]]
+        )
+        self.hazard.get_fraction.return_value = sparse.csr_matrix(
+            [[1.0, 1.0, 1.0], [-0.5, 0.5, 2.0]]
+        )
+        self.exposure_values = np.array([10.0, 20.0, -30.0])
+        self.centroids = np.array([1, 2, 4])
+        self.icalc = ImpactCalc(ENT.exposures, ENT.impact_funcs, self.hazard)
+
+    def test_correct_calculation(self):
+        """Assert that the calculation of the impact matrix is correct"""
+        impact_matrix = self.icalc.impact_matrix(
+            self.exposure_values, self.centroids, ENT.impact_funcs
+        )
+        np.testing.assert_array_equal(
+            impact_matrix.toarray(), [[0.0, 10.0, 30.0], [-5.0, 20.0, -60.0]]
+        )
+
+        # Check if hazard methods were called with expected arguments
+        with self.subTest("Internal call to hazard instance"):
+            self.hazard.get_mdr.assert_called_once_with(
+                self.centroids, ENT.impact_funcs
+            )
+            self.hazard.get_fraction.assert_called_once_with(self.centroids)
+
+    def test_wrong_sizes(self):
+        """Calling 'impact_matrix' with wrongly sized argument results in errors"""
+        centroids = np.array([1, 2, 4, 5])  # Too long
+        with self.assertRaises(ValueError):
+            self.icalc.impact_matrix(self.exposure_values, centroids, ENT.impact_funcs)
+        exposure_values = np.array([1.0])  # Too short
+        with self.assertRaises(ValueError):
+            self.icalc.impact_matrix(exposure_values, self.centroids, ENT.impact_funcs)
+
+
+class TestImpactMatrixGenerator(unittest.TestCase):
+    """Check the impact matrix generator"""
+
+    def setUp(self):
+        """"Initialize mocks"""
+        # Alter the default config to enable chunking
+        self.CONFIG_COPY = deepcopy(CONFIG)
+        CONFIG.max_matrix_size.int = MagicMock(return_value=1)
+
+        # Mock the hazard
+        self.hazard = create_autospec(HAZ)
+        self.hazard.haz_type = "haz_type"
+        self.hazard.centr_exp_col = "centr_col"
+        self.hazard.size = 1
+
+        # Mock the Impact function (set)
+        self.impf = MagicMock(name="impact_function")
+        self.impfset = create_autospec(ENT.impact_funcs)
+        self.impfset.get_func.return_value = self.impf
+
+        # Mock the impact matrix call
+        self.icalc = ImpactCalc(ENT.exposures, self.impfset, self.hazard)
+        self.icalc.impact_matrix = MagicMock()
+
+        # Set up a dummy exposure dataframe
+        self.exp_gdf = pd.DataFrame(
+            {
+                "impact_functions": [0, 11, 11],
+                "centr_col": [0, 10, 20],
+                "value": [0.0, 1.0, 2.0],
+            }
+        )
+
+    def tearDown(self):
+        """Reset the original config"""
+        CONFIG = self.CONFIG_COPY
```

Comment:
Some other tests are currently failing with the error message
```
ValueError: Increase max_matrix_size configuration parameter to > 216
```
I guess this is due to this line not working correctly. I wanted to reset the global `CONFIG` after altering it for this test. Not sure what to do here...

Reply:
@chahank @emanuel-schmid I need some help here. The problem seems to be that I cannot `deepcopy` the `CONFIG` object.

Reply:
I just came up with the solution: Replacing `CONFIG.max_matrix_size` with a new `Config` instance.

The problem is not that `climada.CONFIG` cannot be set (it can), but that the `CONFIG_COPY` still points to the original `CONFIG` that I changed. It's not a real copy.

Reply:
No it doesn't. 😝 Makes sense too. One needs to re-import CONFIG from climada to make it work.

Reply:
See 2a1122f9e18cecb6a8d8d43f8d104f06a0314ddf for the solution (i hope 🙏)

Reply:
Should work (still optimistic). 😄 
However I think deepcopy _does_ work but porting it into the namespace is tricky.

Reply:
It works now, so I'll resolve this discussion. Thanks a lot for your input! 🙌